DOCSP-16466: TLS warning and OCSP stapling (#100)
mongodb/docs-java: source/fundamentals/connection/tls.txt

                })
        .build();
   MongoClient client = MongoClients.create(settings);

Online Certificate Status Protocol (OCSP)
-----------------------------------------

> OCSP is a standard used to check whether X.509 certificates are valid or
> revoked during the TLS handshake. The driver supports the following variations
> of OCSP:

Reviewer: suggestion: I wonder if the removal of the conjunction could simplify this?

```suggestion
OCSP is a standard used to check the validity of X.509 certificates during the TLS handshake. The driver supports the following variations
of OCSP:
```

Author: I think my usage of valid might not be specific enough. I was hoping that the juxtaposition of "valid" and "revoked" would signal opposites. OCSP as far as I know is only to check revocation status of a certificate with information that a revocation server can access. I don't know if revocation servers have any other responsibilities other than checking whether a certificate has been revoked prior to its expiry (whether the certificate is on the Certificate Revocation List), i.e. it doesn't explicitly validate anything about the certificate.
I found some information about the potential statuses that the revocation server can return, so I'll update.

- **Client-Driven OCSP**
- **OCSP Stapling**

The following sections describe the differences between them and how to enable
them for your application.

.. note::

   The Java driver uses the JVM arguments configured for the application
   and cannot be overridden for a specific ``MongoClient`` instance.

Client-Driven OCSP
~~~~~~~~~~~~~~~~~~

In client-driven OCSP, the client sends the certificate in an OCSP request to
an OCSP responder after receiving the certificate from the server. The OCSP
responder checks the status of the certificate with a certificate
authority (CA) and reports whether it's valid in a response sent to the
client.

To enable client-driven OCSP for your application, set the following JVM
system properties:

.. list-table::
   :header-rows: 1
   :widths: 35 65

   * - Property
     - Value

   * - ``com.sun.net.ssl.checkRevocation``
     - Set this property to ``true`` to enable revocation checking.

   * - ``ocsp.enable``

>      - Set this property to ``true`` to enable client-driven OCSP.

Reviewer: Question: Is there any default here, and if there is, would it be worth mentioning? This question applies to the other table in OCSP Stapling as well.

Author: No, there's no default value. The variable names are from the built-in J2SE handling of OCSP:
https://docs.oracle.com/javase/8/docs/technotes/guides/security/jsse/ocsp.html

.. warning::

   If the OCSP responder is unavailable, the TLS support provided by the
   JDK reports a "hard fail". This differs from the "soft fail" behavior of
   the MongoDB Shell and certain drivers.

OCSP Stapling
~~~~~~~~~~~~~

> OCSP Stapling is a type of OCSP in which the server must obtain the signed

Reviewer: ```suggestion
OCSP stapling is a type of OCSP in which the server must obtain the signed
```
suggestion for consistency. The [RFC](https://datatracker.ietf.org/doc/html/rfc6961#section-1) and the  wiki article on it uses lowercase 's'

Author: Good catch!

> certificate from the certificate authority (CA) and including it in a

Reviewer: ```suggestion
certificate from the certificate authority (CA) and include it in a
```

- **Client-Driven OCSP**
- **OCSP Stapling**

The following sections describe the differences between them and how to enable
them for your application.

.. note::

   The Java driver uses the JVM arguments configured for the application
   and cannot be overridden for a specific ``MongoClient`` instance.

Client-Driven OCSP
~~~~~~~~~~~~~~~~~~

In client-driven OCSP, the client sends the certificate in an OCSP request to
an OCSP responder after receiving the certificate from the server. The OCSP
responder checks the status of the certificate with a certificate
authority (CA) and reports whether it's valid in a response sent to the
client.

To enable client-driven OCSP for your application, set the following JVM
system properties:

.. list-table::
   :header-rows: 1
   :widths: 35 65

   * - Property
     - Value

   * - ``com.sun.net.ssl.checkRevocation``
     - Set this property to ``true`` to enable revocation checking.

   * - ``ocsp.enable``
     - Set this property to ``true`` to enable client-driven OCSP.

.. warning::

   If the OCSP responder is unavailable, the TLS support provided by the
   JDK reports a "hard fail". This differs from the "soft fail" behavior of

>    the MongoDB Shell and certain drivers.

Reviewer 2: certain => some other

OCSP Stapling
~~~~~~~~~~~~~

> OCSP stapling is a type of OCSP in which the server must obtain the signed

Reviewer 2: Not sure about `type of OCSP` phrasing.  Is it really a type, or just a mechanism?

Author: I meant "type" as in "variation". I guess it's probably ok to assume the reader can draw the connection with OCSP and describe it instead as "a mechanism in which..."
